### ISSUE TITLE
Poll: Drop Python 2 support?

### ISSUE THREAD

**Issue by richrd** [richrd/suplemon#176]
Hi all! Hopefully this reaches as many people as possible (not only watchers).
TLDR; Vote with thumb up or thumb down reaction.

I'm planning to do some major refactoring and I would like to drop Python 2 support at the same time. I want to solidify the code base and allow for some new features which are currently hard or not possible to implement. Currently some parts of the code (especially input handling) are implemented to support Python 2 and 3 which makes the code more complicated than it should be. The question is, should Python 2 be dropped in favor of better features, easier maintenance and and cleaner code or should Python 2 still be supported?

Here's a list of features that I want do in the future (non exhaustive):
- Support selecting text (regions)
- Per file settings (e.g. different tab width for different file types)
- Split screen view (multiple files visible at the same time)
- Sidebar with directory tree (or anything else that's useful in a sidebar)
- Add a dropdown for autocomplete matches
- Proper plugin architecture (there's a lot that plugins can't do at the moment)
- Proper event system (would really help plugins. for example bracket matching isn't currently possible)

And additionally maybe these features too (non exhaustive):
- Wrapping long lines
- Code folding
- Multi line prompts
- A menu for different commands at the top the screen (with mouse support)

Let me know what you think!

**Comment by twolfson** [comment]
My two cents is to start bundling Python 3 with Suplemon into a standalone executable (Electron apps do this, Sublime Text does this). It will increase the download size but it will remove the worry about someone having a system with the necessary dependencies

**Comment by richrd** [comment]
Not a bad idea!
How should the bundling be done in practice? I know there are tools for it but I haven't looked into it much.
- Does any tool support all platforms (or at least Linux & Mac) 
- Can the bundle be installed or used without root privileges?

**Comment by miohtama** [comment]
What platforms you target? What do you want to accomplish? Do you want to go against Nano/Emacs and other terminal editors or Sublime Text / Atom kind of use cases? 

It looks like you are in the same territory with Nano. So people will mostly use Suplemon a better replacement for in-terminal editing? This would suggest they use Suplemon on a (remote) Linux for terminal editing.

As far as I know today all Linux systems ship with Python 3 installed. Thus, there is no need to support Python 2 or to create separate binary having embedded Python interpreter.

**Comment by richrd** [comment]
@miohtama 
Answers to your questions:
- Platforms: 
    I'm targeting *nix systems, mainly Linux and Mac. I'm not targeting Windows because `The Windows version of Python doesn’t include the curses module.`.
- What I want to accomplish:
    A powerful terminal text editor that is familiar and easy to use. Built in multi cursor support was a core motive to build suplemon. I also thoguht it would be fun and educational to do it (which it is)
- Going against:
    Suplemon is mainly just going against terminal editors, but that doesn't mean it shouldn't have features that desktop editors have. It's mainly rivaling nano (and I dare say it's been beaten by suplemon a while ago, please correct me if I'm wrong). The point of suplemon is to be a powerful terminal editor that feels familiar and is easy to learn (or ultimately you don't need to learn it). However it doesn't need to do everything that heavyweights like vim, emacs, sublime text do.

    One of the most important use cases is editing files over SSH. Here's some background: I got used to using multiple cursors in sublime, and wanted that on the command line too. I know vim & emacs have extensions to do this but they are utterly uninviting to me. I wanted something that you can just run and feel right at home (if you're used to a gui editor like sublime).
- Trying to accomplish:
    - Easier maintenance of code
    - Cleaner code
    - Being able to add more advanced features such as described above

- Platforms: 
    I'm targeting *nix systems, mainly Linux and Mac. I'm not targeting Windows because `The Windows version of Python doesn’t include the curses module.`.
- What I want to accomplish:
    A terminal text ed

- Going against:
    Suplemon is mainly just going against terminal editors, but that doesn't mean it shouldn't have features that desktop editors have. It's mainly rivaling nano (and I dare say it's been beaten by suplemon a while ago, please correct me if I'm wrong). The point of suplemon is to be a powerful terminal editor that feels familiar and is easy to learn (or ultimately you don't need to learn it). However it doesn't need to do everything that eg. vim, emacs, sublime text do.

    One of the most important use cases is editing files over SSH. Here's some background: I got used to using multiple cursors in sublime, and wanted that on the command line too. I know vim & emacs have extensions to do this but they are utterly uninviting to me. I wanted something that you can just run and feel right at home (if you're used to a gui editor like sublime).
- Trying to accomplish:
    - Easier maintenance of code
    - Cleaner code
    - Being able to add more advanced features such as described above


Yes, suplemon shares the same territory as nano, and hopefully more. Hell, suplemon might even be trying to claim nanos territory ;). Editing remote files is definitely one of the most important use cases. I'd like it to be as easy as editing local files. Personally I use suplemon when I need to edit something remotely.

On python versions: Many distros do seem to include python3, but some don't like the latest Centos 7. I sometimes have to work on servers that don't have python3 and I don't have the privileges to install it. If nothing else is available I'm comfortable enough with suplemon 0.1.57.

**Comment by twolfson** [comment]
@richrd I don't know of any tools off the top of my head as I've never implemented cross-platform executable releases (although it's been on my plate for a while in some projects). I'm sure looking at repos like Atom/Electron should provide some modern insight.

I think how you bundle is up to you. I would recommend a wide range hosted by GitHub releases:

- `.deb` (published to `apt` but also can be downloaded to standalone)
- `.rpm` (published to `yum`? but also can be downloaded to standalone)
- `.app` contained in a `.zip` (OS X)
- `.zip`
- `.tar.gz`/`.tar.xz`

Optional support:

- `.msi` (Windows -- if you are bundling, then you can include any packages you want)
- `brew` (it's always nice to easily install/upgrade)

**Comment by indolering** [comment]
@twolfson I think you should migrate your comment to #165.

**Comment by indolering** [comment]
Python 2 is slated for EOL in 2020, I don't think it makes sense to invest time and energy in it.

**Comment by richrd** [comment]
Dropping Python 2 support for new releases.
